### PR TITLE
3273 NDP: Show zero values correctly

### LINKED ISSUE #3273: Urgent(old issue recurring) - zeros do not show up in the main table 1b
@sorja 

**Describe the bug**
See production Bahrain 2000 and 2023 table 1b(see the email)
The NDP page shows the sum result of 0, but not in the main table of 1b

### PR DESCRIPTION
**Pull request by sorja**
- [x] Update ODP view
- [x] Update Dependants

### CODE REVIEW

File: src/test/migrations/steps/20231204133609-step-fix-odp-subfield-calculations.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    // eslint-disable-next-line no-await-in-loop
+    const countryISOs = (await AreaController.getCountries({ assessment, cycle }, client)).map((c) => c.countryIso)
+
+    // eslint-disable-next-line no-await-in-loop
+    const tables = {
+      [TableNames.originalDataPointValue]: {},
+    }
+    const getCountriesDataProps = { assessment, cycle, countryISOs, tables, force: true }
+    // eslint-disable-next-line no-await-in-loop
+    await DataRedisRepository.getCountriesData(getCountriesDataProps, client)
```
Comment (sorja) [inline src/test/migrations/steps/20231204133609-step-fix-odp-subfield-calculations.ts]: Update odp view & cache

File: src/test/migrations/steps/utils/updateODPDependencies.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    const nodes: Array<NodeUpdate> = originalDataPoints.reduce((acc, originalDataPoint) => {
+      const colName = String(originalDataPoint.year)
+      const opdNodes = originalDataPointVariables.map<NodeUpdate>(({ tableName, variableName }) => {
+        return { tableName, variableName, colName, value: undefined }
+      })
+      acc.push(...opdNodes)
+      return acc
+    }, [])
+
```
Comment (minotogna) [inline src/test/migrations/steps/utils/updateODPDependencies.ts]: There's a lot of duplication we could avoid .
Maybe , first generate nodes and then call update dependencies ?

File: src/test/migrations/steps/20231204133609-step-fix-odp-subfield-calculations.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import { getCreateSchemaCycleOriginalDataPointViewDDL } from 'server/repository/assessment/assessment/getCreateSchemaDDL'
+import { DataRedisRepository } from 'server/repository/redis/data'
+
+import { updateODPDependencies } from 'test/migrations/steps/utils/updateODPDependencies'
+
+export default async (client: BaseProtocol) => {
+  const assessment = await AssessmentController.getOne({ assessmentName: 'fra', metaCache: true }, client)
+
+  // // eslint-disable-next-line no-restricted-syntax
+  for (let i = 0; i < assessment.cycles.length; i += 1) {
```
Comment (minotogna) [inline src/test/migrations/steps/20231204133609-step-fix-odp-subfield-calculations.ts]: Can we use promises each ? So we avoid all eslint disable ?

File: src/test/migrations/steps/utils/updateODPDependencies.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  const countryNodes: Record<CountryIso, Array<NodeUpdate>> = {} as Record<CountryIso, Array<NodeUpdate>>
+
+  const originalDataPointVariables = getOriginalDataPointVariables({ cycle })
+
+  const schemaCycle = Schemas.getNameCycle(assessment, cycle)
+  const originalDataPoints = await client.map<{ countryIso: CountryIso; year: number }>(
+    `
+        select country_iso, year
+        from ${schemaCycle}.original_data_point
+        where national_classes != '[]'::jsonb
```
Comment (minotogna) [inline src/test/migrations/steps/utils/updateODPDependencies.ts]: I think there's a jsonb method that returns the length of an array or even empty . Would that be safer ?

File: src/test/migrations/steps/utils/updateODPDependencies.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  `,
+    [],
+    (res) => Objects.camelize(res)
+  )
+
+  await Promises.each(originalDataPoints, async ({ countryIso, year }) => {
```
Comment (minotogna) [inline src/test/migrations/steps/utils/updateODPDependencies.ts]: I don't think there's a need to use promises each here .

File: src/tools/getDeps.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import { AssessmentMetaCaches } from 'meta/assessment'
+
+import { AssessmentController } from 'server/controller/assessment'
+import { Logger } from 'server/utils/logger'
+
+const exec = async () => {
+  if (process.argv.length === 6) {
+    const [assessmentName, cycleName, tableName, variableName] = process.argv.slice(2)
+
+    const { assessment, cycle } = await AssessmentController.getOneWithCycle({
```
Comment (minotogna) [inline src/tools/getDeps.ts]: I don't think we need this . Dependencies and dependents can be seen in redux store or database

Comment (sorja) [reply src/tools/getDeps.ts]: Sure, I use it locally when working with this kind of tasks

**Review by minotogna — APPROVED** [review]

**Pull request merged**